Use existing joints to maintain balance among processes

shellshocked2003 opened the pull request:
Selects joints from input joint sets that divide initial rock volume into approximately equal volumes, depending on number of processors that will be used in analysis. This eliminates the need to use either `treeReduce` or `groupBy`, both of which are very expensive since they cause data to be shuffled across the network. Changes include:
* Change input file format to be much simpler and smaller. Joint sets are generated within `sparkRocks` by new `JointGenerator` class; user only needs to provide parameters that are obtained from field investigation
* Eliminated need for `LoadBalancer` class since only real joints are used in maintaining load balance
* Addresses #35 since this is no longer needed
* Addresses most of the concerns in #10
* #31 should no longer be of concern since this was artificially introduced by processor joints

Stochastic and non-persistent joint generators have place holders, but have not yet been implemented. These will be implemented at a later date in a different pull request.

jhkolb commented on src/main/scala/edu/berkeley/ce/rockslicing/InputProcessor.scala:
```diff
@@ -1,16 +1,15 @@
-  def readInput(inputSource: Source): Option[(Array[Double],
-                                             Seq[Face], Seq[Joint])] = {
+  def readInput(inputSource: Source): Option[ (Array[Double], Array[Double],
+    Vector[Array[Double]], Vector[Array[Double]]) ] = {
```
Can we use `Seq` here for the last types?

The caller of this function shouldn't care if they get back a `Vector` specifically.

shellshocked2003 replied:
Yeah, can't remember why `Vector` was specified, but it's gone now.

jhkolb commented on src/main/scala/edu/berkeley/ce/rockslicing/InputProcessor.scala:
```diff
@@ -1,16 +1,15 @@
+    Vector[Array[Double]], Vector[Array[Double]]) ] = {
     val lines = inputSource.getLines().zipWithIndex.toVector
     val globalOriginLine = lines.head._1
+    val boundingBoxLine = lines.tail.head._1
```
I would suggest we write `lines(1)._1` instead, as that is much simpler. I think chained calls to `head` and `tail` are usually best avoided.

jhkolb commented on src/main/scala/edu/berkeley/ce/rockslicing/InputProcessor.scala:
```diff
@@ -24,10 +23,23 @@ object InputProcessor {
 
-    val remainingLines = lines.tail
-    val transitionIndex = remainingLines.indexWhere(_._1 == "%")
+    val boundingBoxTokens = boundingBoxLine.split(" ") map {x => Try(x.toDouble) }
```
For consistent formatting, you need a space before the `x`, i.e. `map {x ...` should become `map { x ...`.

jhkolb commented on src/main/scala/edu/berkeley/ce/rockslicing/InputProcessor.scala:
```diff
@@ -24,10 +23,23 @@ object InputProcessor {
+    val boundingBoxArr = Array(boundingBox(0), boundingBox(1), boundingBox(2),
+                               boundingBox(3), boundingBox(4), boundingBox(5))
+
+    val remainingLines = lines.tail.tail
```
This is also weird. Let's just do a more idiomatic `lines.drop(2)`

shellshocked2003 replied:
These weird things seem to make sense in the thick of it... When looking at it again it definitely looks weird.

jhkolb commented on src/main/scala/edu/berkeley/ce/rockslicing/InputProcessor.scala:
```diff
@@ -42,33 +54,14 @@ object InputProcessor {
-      if (doubleVals.length != 6) {
-        println(s"Error, Line $index: Each face of input rock volume is defined by values: a, b, c, d, phi and cohesion")
+      if (doubleVals.length != 7) {
+        println(s"Error, Line $index: Each face of input rock volume is defined by values: strike, dip, x, y, z,"+
```
How did you decide on the order here?

shellshocked2003 replied:
You mean the order of the inputs? Figured geometric inputs first, followed by surficial parameters. Does that make sense?

jhkolb replied:
Yeah, that sounds good.

jhkolb commented on src/main/scala/edu/berkeley/ce/rockslicing/InputProcessor.scala:
```diff
@@ -79,45 +72,31 @@ object InputProcessor {
-      val (mandatoryValues, optionalValues) = doubleVals.splitAt(11)
-      if (optionalValues.length % 4 != 0) {
-        println(s"Error, Line $index: Bounding Faces for Block Require 4 Values Each")
+      val optionalValues = doubleVals.drop(6)
```
I think you could simplify this by getting rid of the `optionalValues` val and then checking the length of `doubleVals` against 6, 9, and 10 rather than checking the length of `optionalValues` against 0, 3 and 4.

shellshocked2003 replied:
Good point. I'll change this.

jhkolb commented on src/main/scala/edu/berkeley/ce/rockslicing/Joint.scala:
```diff
@@ -13,8 +13,8 @@ object Joint {
-  private def findDistance(normalVec: Array[Double], localOrigin: Array[Double],
-                            center: Array[Double]): Double = {
+  def findDistance(normalVec: Array[Double], localOrigin: Array[Double],
+                           center: Array[Double]): Double = {
```
Can you change the indentation here back to its original form? `center` should line up with `normalVec`.

jhkolb commented on src/main/scala/edu/berkeley/ce/rockslicing/JointGenerator.scala:
```diff
@@ -0,0 +1,221 @@
+    val dipRadians = dip.toRadians
+    val strikeVector = DenseVector[Double](math.cos(-strikeRadians), math.sin(-strikeRadians), 0.0)
+    // If joint is vertical, dip vector will point along negative z-axis
+    if (math.abs(dip - 90.0) < NumericUtils.EPSILON) {
```
How about you do something more like

``` scala
val dipVector = if (math.abs(dip - 90.0) < NumericUtils.EPSILON) {
    DenseVector[Double](0.0, 0.0, -1.0)
} else {
    DenseVector[Double](math.cos(dipRadians)...
}
val jointNormal = ...
Array(...
```

This allows you to avoid a bit of duplicate code.

shellshocked2003 replied:
Yep, this is better.

jhkolb commented on src/main/scala/edu/berkeley/ce/rockslicing/JointGenerator.scala:
```diff
@@ -0,0 +1,221 @@
+          phi=parameters(4), cohesion=parameters(5), shape=Vector.empty))
+      } else {
+        // TODO: Implement stochastic generator
+        println("Error, Joint Input Data: Stochastic joint generation not implemented yet")
```
The standard thing to do in this kind of situation is to throw an `UnsupportedOperationException`.

This way, if someone tries to use this feature, the program will stop immediately and the user can look at the stack trace to get a sense of what's going on.

jhkolb commented on src/main/scala/edu/berkeley/ce/rockslicing/JointGenerator.scala:
```diff
@@ -0,0 +1,221 @@
+    *
+    * @param globalOrigin Coordinates of the global origin
+    * @param lowerLeftCorner Lower left corner of the bounding box
+    * @param joints Array containing the input data representing joint sets
```
So should this parameter really be named `jointSets`?

jhkolb commented on src/main/scala/edu/berkeley/ce/rockslicing/JointGenerator.scala:
```diff
@@ -0,0 +1,221 @@
+        None
+      }
+    }
+    if (masterJoints.length != joints.length) {
```
I'm not really clear on why an option is used here.

Where this function is called, it just performs a `get`.

Is the user reasonably expected to be able to recover from a situation where this function returns `None` and keep running, or should we just throw an exception to simplify things?

shellshocked2003 replied:
The option was there to catch if the user had input parameters for stochastic joint generation, but that's taken care of by an `UnsupportedOpertationException` now so I'll get rid of the option on this.

jhkolb commented on src/main/scala/edu/berkeley/ce/rockslicing/JointGenerator.scala:
```diff
@@ -0,0 +1,221 @@
+  * @param globalOrigin Coordinates of the global origin
+  * @param boundingBox Coordinates specifying the lower left and upper right corners of the box that bounds
+  *                    the domain of interest.
+  * @param rockVolumeData Array of arrays describing each of the faces that define the boundaries of
```
I think this should read "Seq of Arrays" instead of "Array of Arrays."

jhkolb commented on src/main/scala/edu/berkeley/ce/rockslicing/JointGenerator.scala:
```diff
@@ -0,0 +1,221 @@
+  *                       the rock volume of interest. The inputs in each array are strike, dip and the x,
+  *                       y and z coordinates of a point that lies within the plane containing the face.
+  *                       Strike and dip should be specified in degrees.
+  * @param jointSetData Array of arrays containing the input data representing joint sets. The inputs in each array
```
Also "Seq of  Arrays".

jhkolb commented on src/main/scala/edu/berkeley/ce/rockslicing/JointGenerator.scala:
```diff
@@ -0,0 +1,221 @@
+    * @return Seq of joints for each input joint set
+    */
+  def generateJointSets(jointSetData: Seq[Array[Double]]): Seq[Seq[Joint]] = {
+    val jointTuple = masterJoints.zip(jointSetData)
```
Since this is a collection, how about the name `jointTuples` instead of `jointTuple`?

jhkolb commented on src/main/scala/edu/berkeley/ce/rockslicing/JointGenerator.scala:
```diff
@@ -0,0 +1,221 @@
+  }
+
+  /**
+    * Finds representative joint for each joint set, centered at the lower left corner of the
```
Can you add some comments making it more clear what it means for a joint to be a representative joint?

jhkolb commented on src/main/scala/edu/berkeley/ce/rockslicing/JointGenerator.scala:
```diff
@@ -0,0 +1,221 @@
+  *                     are strike, dip, joint spacing, persistence, phi, cohesion and optional stochastic parameters
+  */
+case class JointGenerator(globalOrigin: Array[Double], boundingBox: Array[Double], rockVolumeData: Seq[Array[Double]],
+                     jointSetData: Seq[Array[Double]]) {
```
How about lining up `jointSetData` with `globalOrigin` on the indentation?

jhkolb commented on src/main/scala/edu/berkeley/ce/rockslicing/JointGenerator.scala:
```diff
@@ -0,0 +1,221 @@
+  val masterJoints = JointGenerator.findMasterJoints(globalOrigin, lowerLeftCorner, jointSetData).get
+
+  if (masterJoints.isEmpty) {
+    throw new IllegalArgumentException("ERROR: Stochastic joint generation not yet implemented")
```
This is exactly like what I had in mind above. `IllegalArgumentException` works but `UnsupportedOperationException` might be a bit more standard.

shellshocked2003 replied:
Moved this inside `findMasterJoints` so there's no need to throw the exception here anymore since it'll be caught earlier.

jhkolb commented on src/main/scala/edu/berkeley/ce/rockslicing/JointGenerator.scala:
```diff
@@ -0,0 +1,221 @@
+    }
+
+    jointTuple map { case (joint, jointData) =>
+      if (jointData(3) == 100) {
```
All these arrays to represent complex data types are confusing. :)

I have some ideas on how we can improve this I'll add in a comment on the overall pull request.

jhkolb commented on src/main/scala/edu/berkeley/ce/rockslicing/JointGenerator.scala:
```diff
@@ -0,0 +1,221 @@
+    */
+  def makeNonPersistentJointSet(joint: Joint, jointData: Array[Double]): Seq[Joint] = {
+    // Still needs to be implemented, but will require stochastic generator
+    println("ERROR: Non-persistent joint generator not yet implemented")
```
Let's throw an `UnsupportedOperationException` here.

jhkolb commented on src/main/scala/edu/berkeley/ce/rockslicing/JointGenerator.scala:
```diff
@@ -0,0 +1,221 @@
+  @tailrec
+  private def makePersistentJointSet(joint: Joint, jointData: Array[Double], diagonalVector: DenseVector[Double],
+                                     center: Array[Double], terminationPoint: Array[Double],
+                                     jointSet: Seq[Joint]): Seq[Joint] = {
```
How about giving `jointSet` a default value of `Seq.empty[Joint]` so that we can omit it on the initial call?

shellshocked2003 replied:
Done!

jhkolb commented on src/main/scala/edu/berkeley/ce/rockslicing/JointGenerator.scala:
```diff
@@ -0,0 +1,221 @@
+    *                 first called.
+    * @return Seq of joints representing the joint set
+    */
+  @tailrec
```
I like the tail recursion and I think this function is pretty clear. Doesn't seem like a `fold` would apply here like we were discussing.

shellshocked2003 replied:
Cool. Yeah, I was trying to think about how to do it with `fold`, but it didn't seem like it would work as naturally as the recursion.

jhkolb commented on src/main/scala/edu/berkeley/ce/rockslicing/RockSlicer.scala:
```diff
@@ -33,19 +32,44 @@ object RockSlicer {
-      processorJoints.foldLeft(starterBlocks) { (currentBlocks, joint) =>
-        currentBlocks.flatMap(_.cut(joint))
+    val seedJoints = if (arguments.numProcessors > 1) {
+      // Check if at least on of the input joint sets is persistent
```
Should be "at least one of the input joint sets is persistent"

jhkolb commented on src/main/scala/edu/berkeley/ce/rockslicing/RockSlicer.scala:
```diff
@@ -33,19 +32,44 @@ object RockSlicer {
+        jointSet(3) == 100
+      }
+
+      if (!persistentCheck) {
```
Don't save this as a separate `val`. Instead, just do

``` scala
if (!jointSetInputs.exists(_(3) == 100)
```

or

``` scala
if (jointSetInputs.forall(_(3) != 100))
```

This code will also become clearer if we follow my advice about representing joint sets.

jhkolb commented on src/main/scala/edu/berkeley/ce/rockslicing/RockSlicer.scala:
```diff
@@ -33,19 +32,44 @@ object RockSlicer {
+        System.exit(-1)
+      }
+
+      val seedJoints = SeedJointSelector.searchJointSets(generatedInput.jointSets,
```
How about changing this function to return both the seed joints and the remaining joints?

That would get rid of the `jointSets.flatten.diff` logic below, which I thought was a bit messy.

shellshocked2003 replied:
Okay, I've changed this. Take a look at `searchJointSets` and see what you think. Still need to do a `diff` on one of the joint sets, but the others are just flattened. Don't think there's a way round that.

jhkolb commented on src/main/scala/edu/berkeley/ce/rockslicing/RockSlicer.scala:
```diff
@@ -33,19 +32,44 @@ object RockSlicer {
+        println("ERROR: Unable to find satisfactory seed joints")
+        System.exit(-1)
       }
+      seedJoints.get
```
Hmm, so `searchJointSets` returns an option?

That seems kind of silly when you just call `get` on it later. Maybe you should consider just throwing some kind of error rather than returning `None`?

shellshocked2003 replied:
I've changed this to throw and exception earlier. There's definitely some abuse of returning `None` and catching it later instead of throwing an exception immediately, trying to eradicate that throughout.

jhkolb commented on src/main/scala/edu/berkeley/ce/rockslicing/SeedJointSelector.scala:
```diff
@@ -0,0 +1,243 @@
+  def searchJointSets(jointSets: Seq[Seq[Joint]], inputVolume: Block,
+                      numProcessors: Int): Option[Seq[Joint]] = {
+    if (jointSets.isEmpty) {
+      println("ERROR: Unable to find seed joints")
```
I'm leaning towards throwing an exception here. What do you think?

My sense is that returning an option and then just calling `get` on it isn't the best style. To me, returning an option means that the function could reasonably return `None` or `Some` and the caller will continue normal operation regardless, which isn't true in our case as far as I can tell.

shellshocked2003 replied:
Agreed. See previous comment.

jhkolb commented on src/main/scala/edu/berkeley/ce/rockslicing/SeedJointSelector.scala:
```diff
@@ -0,0 +1,243 @@
+      if (seedJointOption.isEmpty) {
+          searchJointSets(jointSets.tail, inputVolume, numProcessors)
+      } else {
+        seedJointOption
```
Don't we know that `seedJointOption` is `None` here anyways, so we could just return that explicitly (if we don't opt for some kind of exception-based approach).

shellshocked2003 replied:
Here `seedJointOption` won't be `None` necessarily. This is the case when it's `Some` and we can stop the recursion and return. When it is `None`, we move on to the next joint set to check if that one yields satisfactory seed joints. 
I've modified the code to throw an exception instead of returning `None`. I think your suggestion to do this makes the code much cleaner and clearer.

jhkolb commented on src/main/scala/edu/berkeley/ce/rockslicing/SeedJointSelector.scala:
```diff
@@ -0,0 +1,243 @@
+      // Check that joint set is persistent
+      val seedJointOption = findSeedJoints(jointSets.head, inputVolume, numProcessors, inputVolume.volume)
+
+      if (seedJointOption.isEmpty) {
```
Oh okay so we just keep trying each non-persistent joint set until we get one that works, right?

Can you add some comments to clarify this?

shellshocked2003 replied:
Done

jhkolb commented on src/main/scala/edu/berkeley/ce/rockslicing/SeedJointSelector.scala:
```diff
@@ -0,0 +1,243 @@
+    * @param rockVolume Block representing input rock volume that should be divided into approximately equal pieces
+    *                   for each process to operate on
+    * @param numProcessors Number of processes that will be used in analysis
+    * @param totalVolume Total volume of rockVolume
```
Can't we just calculate this within the function rather than asking for it as input?

The fewer the parameters to this function, which already has a few, the better.

shellshocked2003 replied:
Yeah, we can do that. I just passed it as a parameter so it wouldn't have to recalculate the same number over and over again since the total volume won't change.

jhkolb commented on src/main/scala/edu/berkeley/ce/rockslicing/SeedJointSelector.scala:
```diff
@@ -0,0 +1,243 @@
+    *                  will be updated as necessary in recursive calls.
+    * @param stepSize Step size that tolerance will be incremented by if unable to find enough seed joints. Default
+    *                 initial value is 1%, though it will be updated as necessary in recursive calls.
+    * @return If able to find the required number of seed joints, this function will return a Seq of Joints. If unable
```
So if I'm understanding correctly, we might call this function multiple times if the first few calls fail, right? This seems like a better place for an option.

shellshocked2003 replied:
Yeah, you're understanding correctly. On each call, the function will either return the required number of seed joints to approximately balance the load. It it is unable to find the required number of seed joints, it returns `None` at which point the caller moves on to the next joint set.
I think I was using the `Some`/`None` idea in place of throwing exceptions in parts of the new code. I'll fix those.

jhkolb commented on src/main/scala/edu/berkeley/ce/rockslicing/SeedJointSelector.scala:
```diff
@@ -0,0 +1,243 @@
+                     tolerance: Double = 0.01, stepSize: Double = 0.01): Option[Seq[Joint]] = {
+    if (jointSet.length < numProcessors - 1) {
+      println(s"Error: Not enough joints to generate required number of seed joints. ${jointSet.length} joints but"+
+      s" $numProcessors processors. Try using less processors.")
```
Well since I've already been picky about code in this review, why not add English to the mix? I think it would be better to say "Try using fewer processors". This is really not that important though :)

shellshocked2003 replied:
This is very important, should have fixed it first :)

jhkolb commented on src/main/scala/edu/berkeley/ce/rockslicing/SeedJointSelector.scala:
```diff
@@ -0,0 +1,243 @@
+      Seq.empty[Joint], tolerance, numProcessors)
+
+    if (stepSize < 0.01) {
+      println("Error: Unable to find satisfactory seed joints using current joint set. Please try another " +
```
Is it actually useful to print out this message? I thought that, as our implementation stands now, the user can't control which persistent joint set gets tried, so maybe they shouldn't be presented with a message like this. We handle it and try as many as we have to until we get a joint set that works, right?

shellshocked2003 replied:
Yeah, I'll take this out. I think this was more in there for me while I was debugging.

jhkolb commented on src/main/scala/edu/berkeley/ce/rockslicing/SeedJointSelector.scala:
```diff
@@ -0,0 +1,243 @@
+      val nonRedundantBlocks = blocks map { case block @ Block(blockCenter, _, _) =>
+          Block(blockCenter, block.nonRedundantFaces)
+      }
+      val volumes = nonRedundantBlocks.map(_.volume)
```
How about rewriting this as:

``` scala
val avgVolume = volumes.sum / volumes.length
val maxImbalance = nonRedundantBlocks.map(block => math.abs(block.volume - avgVolume)).max
```

Maybe I'm missing something, but this gets rid of the conditional.

shellshocked2003 replied:
I think it does, but this is leftover code that should have been erased. It's gone now, unconditionally :)

jhkolb replied:
Yeah, I think the new code looks better.

Also, do we still have the block's volume as a method? We could consider using a `lazy val` instead if a block's volume needs to be reused in several places.

shellshocked2003 replied:
Yeah, we still have it as a method. The only case where it gets called multiple times on the same block is when searching the different joint sets; in all other cases it's called only after all blocks have been cut and the centroid etc. needs to be calculated for output. 
So, for a standard input it may be called perhaps one to five times? I'm not sure how much difference it would make to have it as a method vs `lazy val`. What do you think would be best?

jhkolb replied:
I think it depends on how expensive the volume computations are. If they're pretty involved, we could consider using a lazy val. If it involves an LP stuff I'd say it could be worth it.

But this could all be premature optimization anyways.

shellshocked2003 replied:
The change is pretty minor code-wise, so just changed it.

jhkolb replied:
Nice. That was exactly the way I would have changed it too.

jhkolb commented on src/main/scala/edu/berkeley/ce/rockslicing/SeedJointSelector.scala:
```diff
@@ -0,0 +1,243 @@
+      } else {
+        math.abs(volumes.min - avgVolume) / avgVolume
+      }
+      println("\nLOAD BALANCE INFORMATION:")
```
What's your opinion on removing this message? It's useful for debugging obviously, but do you think we should still present it to the user? I could see arguments for keeping it or removing it.

shellshocked2003 replied:
I think we should leave this in there since it'll be helpful to users to decide what's the best number of processors to use - the load balance is ad hoc and checking balance using +/- a processor or two will help pick the best number in terms of balancing volumes across processors.

jhkolb replied:
Okay, let's keep it then.

jhkolb commented on src/main/scala/edu/berkeley/ce/rockslicing/SeedJointSelector.scala:
```diff
@@ -0,0 +1,243 @@
+    * @return
+    */
+  @tailrec
+  private def cycleJointSet(jointSet: Seq[Joint], initialBlock: Block, volumePerProcessor: Double,
```
The use of `cycle` is confusing. That will probably make people think either of cycles in graphs or cycling as in shifting the elements of a sequence and wrapping around.

What does this function do and how can we name it accordingly?

shellshocked2003 replied:
The function iterates through the input joint set and finds a `Seq` of joints that divides the input block into approximately equal volumes - these will be used as the seed joints. Changed the name to `selectSeedJoints`, which states more clearly what it actually does.

jhkolb commented on src/main/scala/edu/berkeley/ce/rockslicing/SeedJointSelector.scala:
```diff
@@ -0,0 +1,243 @@
+    val blockSet1 = inputBlock cut joint1
+    val blockSet2 = inputBlock cut joint2
+
+    val nonRedundantBlocks = blockSet1 map { case block @ Block(blockCenter, _, _) =>
```
We may want to name this `nonRedundantBlocks1` for consistency with the other `val`s in this function.

jhkolb commented on src/main/scala/edu/berkeley/ce/rockslicing/SeedJointSelector.scala:
```diff
@@ -0,0 +1,243 @@
+    * @return Returns tuple containing joint that provides best division of volume and remaining volume that is to be
+    *         subdivided further. If neither joint yields acceptable subdivision, function returns None.
+    */
+  private def testVolumes(joint1: Joint, joint2: Joint, inputBlock: Block,
```
I have to admit I found this function pretty confusing.

Why do we consider joints two at a time instead of one at a time? May want to add some comments explaining this.

shellshocked2003 replied:
Added some more comments to the function documentation that should explain what the function is doing. Let me know that helps make it clearer. The function body is pretty heavily commented, but hopefully the "big picture" description in the documentation makes it clearer. 
The first portion of the conditionals deals with cases when one of the input joints doesn't intersect the block - this is either near the beginning or end of the joint set when joints are near the vertices of the bounding box. The final `else` in the function is where both joints intersect the volume and most of the inputs will fall. Does this make sense?

jhkolb replied:
Yeah, the comments make things clearer after a quick glance. I'll take a closer look when I do my second pass.

jhkolb commented on src/main/scala/edu/berkeley/ce/rockslicing/SeedJointSelector.scala:
```diff
@@ -0,0 +1,243 @@
+      val volumeDiff2 = (sortedBlocks2(0).volume - desiredVolume) / desiredVolume
+
+      // Return joint that gives volume closest to desired volume
+      // When less than 50% of the total volume remains, begin loosening constaints on joint selected until
```
What determined the choice of 50% as the threshold? Is it just arbitrary? If so, we may want to extract this out as a constant.

shellshocked2003 replied:
Pretty arbitrary, so changed it to a constant

jhkolb commented:
Okay, finished a first pass. Did you know that this pull request deletes more lines than it adds? So we'll have a net decrease in code size, which is pretty nice.

The one big thing that stood out to me is that we're using `Array`s in places to represent some more complex data types, and that's really making the code harder to understand.

The main example is using an array of 7 doubles to represent a joint set (if I'm remembering stuff I was seeing at the beginning of my review correctly). If this were Python, we'd use a named tuple to make the code clearer, but as we're in Scala I'd say we should go ahead and use a case class. So we'd have a new `JointSet` case class with strike, dip, x, y, z, etc. members and we'd generate these when processing the input file. Plus, now we can use pattern matching to extract joint set fields as we're iterating through collections of joint sets for cleaner code.

Yes this will add some memory overhead, but it'll only be incurred when we're trying to do the initial load balancing, and memory hasn't been an issue during this phase of execution so far.

So now, we won't have something like `jointSet(3) == 100` but instead have `jointSet.Persistence == 100` or whatever we decide to name that field. Seems a lot better to me!

shellshocked2003 commented:
Added a new case class `JointSet` like you suggested and it definitely makes the code more readable. Take a look and see what you think. The rock volume inputs are also quite long, but not as long as the joint inputs. Since we don't do nearly as many operations on those, I figured it would be overkill to make another case class for that. Do you agree with that or should I add something for that too?

jhkolb commented on src/main/scala/edu/berkeley/ce/rockslicing/InputProcessor.scala:
```diff
@@ -1,16 +1,15 @@
   // Processes input file: Add rock volume faces and joints to respective input list
-  def readInput(inputSource: Source): Option[(Array[Double],
-                                             Seq[Face], Seq[Joint])] = {
+  def readInput(inputSource: Source): Option[ (Array[Double], Array[Double],
```
Can we add full JavaDoc style documentation for this function, since it is visible and used elsewhere in our code?

shellshocked2003 replied:
For you, anything!

jhkolb commented on src/main/scala/edu/berkeley/ce/rockslicing/InputProcessor.scala:
```diff
@@ -79,45 +72,27 @@ object InputProcessor {
-                        3 For Local Origin
-                        3 For Joint Center
+      if (doubleVals.length < 6) {
+        println(s"""Error, Line $index: Each input joint is defined by at least 6 values:
```
Shouldn't this now read something like `Each input join set is defined....`?

shellshocked2003 replied:
Yeah, good catch.

jhkolb commented on src/main/scala/edu/berkeley/ce/rockslicing/InputProcessor.scala:
```diff
@@ -1,16 +1,15 @@
-  def readInput(inputSource: Source): Option[(Array[Double],
-                                             Seq[Face], Seq[Joint])] = {
+  def readInput(inputSource: Source): Option[ (Array[Double], Array[Double],
+    Seq[Array[Double]], Seq[JointSet]) ] = {
```
I'm leaning towards us using more abstraction for the rock volume as well. Is there any reason we can't just use our `Face` class, and return a `Seq` of `Face` instances? If I'm not mistaken, it should already have all of the necessary fields, right?

shellshocked2003 replied:
I'll add a case class for this. `Face` won't work since the inputs also contain a point in the plane that defines the face - this isn't stored in `Face` since everything describing the face is relative to some local origin.

jhkolb commented on src/main/scala/edu/berkeley/ce/rockslicing/JointGenerator.scala:
```diff
@@ -0,0 +1,210 @@
+    */
+  private def findBoundingFaces(globalOrigin: Array[Double], rockVolumeData: Seq[Array[Double]]): Seq[Face] = {
+    rockVolumeData map { parameters =>
+      val faceCenter = Array(parameters(2), parameters(3), parameters(4))
```
Yeah, seeing this code makes me think we should express the rock volume as a collection of `Face` instances. This code would become _much_ more readable. For example:

``` scala
rockVolume map { case Face(strike, dip, center) =>
    // You don't even need an explicit faceCenter any more
    val normal = findJointNormal(strike, dip)
```

But this chunk of code also reminded me that our `Face` objects are slightly different in that they are defined by normal vector and distance rather than center. So maybe we need a separate (and very simple) case class for the input faces?

I just think we should use the benefits of higher level programming languages, i.e. giving human readable names to things, where they make for clearer code and don't cost us much.

Also, this can help get rid of a lot of those `Array[Array[Double]]` instances in the unit tests.

shellshocked2003 replied:
Yeah, I'll add this. Should be pretty simple.

jhkolb commented on src/main/scala/edu/berkeley/ce/rockslicing/JointGenerator.scala:
```diff
@@ -0,0 +1,210 @@
+  * @param jointSetData Seq of arrays containing the input data representing joint sets. The inputs in each array
+  *                     are strike, dip, joint spacing, persistence, phi, cohesion and optional stochastic parameters
+  */
+case class JointGenerator(globalOrigin: Array[Double], boundingBox: Array[Double], rockVolumeData: Seq[Array[Double]],
```
So now `rockVolumeData` could become a `Seq` of our new case class.

jhkolb replied:
I'm also wondering if we don't want to make `boundingBox` a pair: `(Array[Double], Array[Double])` with each element representing one of the corners. It's a little complicated but might make the separation clearer. It's up to you.

shellshocked2003 replied:
Like I said Jack, for you, anything.

jhkolb commented on src/main/scala/edu/berkeley/ce/rockslicing/JointGenerator.scala:
```diff
@@ -0,0 +1,210 @@
+    *                     are strike, dip, joint spacing, persistence, phi, cohesion and optional stochastic parameters
+    * @return Seq of joints for each input joint set
+    */
+  def generateJointSets(jointSetData: Seq[JointSet]): Seq[Seq[Joint]] = {
```
It appears that the only place this method is called is a few lines above. Is there any reason not to make it `private` since it's only used within this class?

Also, let's change the name `jointSetData` to `jointSets`. All variables are already data in some sense, right? :)

jhkolb commented on src/main/scala/edu/berkeley/ce/rockslicing/JointGenerator.scala:
```diff
@@ -0,0 +1,210 @@
+    * @return Seq of joints representing the joint set
+    */
+  @tailrec
+  private def makePersistentJointSet(joint: Joint, jointData: JointSet, diagonalVector: DenseVector[Double],
```
How about just `jointSet: JointSet`?

shellshocked2003 replied:
Made it `jointProperties` since there's already another `jointSet`

jhkolb commented on src/main/scala/edu/berkeley/ce/rockslicing/JointSet.scala:
```diff
@@ -0,0 +1,45 @@
+  * @param parameters Array containing the input data representing the joint set. The inputs in each array
+  *                   are strike, dip, joint spacing, persistence, phi, cohesion and optional stochastic parameters
+  */
+case class JointSet(parameters: Array[Double]) {
```
Remember that constructors are a little different for case classes. They determine how we can do pattern matching. So this really needs to be

``` scala
case class JointSet(strike: Double, dip: Double, jointSpacing: Double, persistence: Double, phi: Double, cohesion: Double)
```

because it lets us do cool stuff like this:

``` scala
jointSets map { case JointSet(strike, dip, jointSpacing, persistence, _, _) =>
    ...
}
```

Okay, I see that there's also standard deviation. We can further simplify things by including those as parameters with default values:

``` scala
case class JointSet(strike: Double, jointSpacing: Double, ... strikeStdDev: Double = 0, dipStdDev: Double = 0, ...)
```

jhkolb commented on src/main/scala/edu/berkeley/ce/rockslicing/JointSet.scala:
```diff
@@ -0,0 +1,45 @@
+  val cohesion = parameters(5)
+
+  // Indicates whether joint set is deterministic or stochastic
+  val stochasticFlag = if (parameters.length > 6) {
```
Let's change the name to `isStochastic`. And we can simplify things like this:

``` scala
val isStochastic = parameters.length > 6
```

jhkolb commented on src/main/scala/edu/berkeley/ce/rockslicing/JointSet.scala:
```diff
@@ -0,0 +1,45 @@
+    false
+  }
+
+  // Optional stochastic parameters
```
This logic will all go away if we include these as constructor parameters with default values of `0.0`.

jhkolb commented on src/main/scala/edu/berkeley/ce/rockslicing/RockSlicer.scala:
```diff
@@ -33,21 +32,38 @@ object RockSlicer {
+        jointSet.persistence == 100
+      }
+
+      if (!persistentCheck) {
```
Let's avoid declaring a `val` separately like this and change the code to something more like

``` scala
if (jointSetInputs.forAll(_.persistence != 100))
```

shellshocked2003 replied:
Yeah, this is much cleaner.

jhkolb commented on src/main/scala/edu/berkeley/ce/rockslicing/SeedJointSelector.scala:
```diff
@@ -0,0 +1,257 @@
+    *         joint set.
+    */
+  @tailrec
+  def searchJointSets(jointSets: Seq[Seq[Joint]], inputVolume: Block,
```
Very picky, but can you fix up the indentation here?

shellshocked2003 replied:
Yeah, fixed it. That's actually what Intellij thinks is appropriate...

jhkolb commented on src/main/scala/edu/berkeley/ce/rockslicing/SeedJointSelector.scala:
```diff
@@ -0,0 +1,257 @@
+  private def findSeedJoints(jointSet: Seq[Joint], rockVolume: Block, numProcessors: Integer, tolerance: Double = 0.01,
+                             stepSize: Double = 0.01): Option[Seq[Joint]] = {
+    if (jointSet.length < numProcessors - 1) {
+      throw new IllegalArgumentException(s"Error: Not enough joints to generate required number of seed joints. ${jointSet.length} joints but"+
```
Do you think it's appropriate to throw an exception rather than return `None` in this case? Couldn't this function just be called again with a different joint set?

shellshocked2003 replied:
You're right, I'll change this.

jhkolb commented on src/main/scala/edu/berkeley/ce/rockslicing/SeedJointSelector.scala:
```diff
@@ -0,0 +1,257 @@
+    */
+  @tailrec
+  private def selectSeedJoints(jointSet: Seq[Joint], initialBlock: Block, volumePerProcessor: Double,
+                            totalVolume: Double, selectedJoints: Seq[Joint], tolerance: Double,
```
Let's fix this up so that the first parameter on each line is vertically aligned.

jhkolb commented on src/main/scala/edu/berkeley/ce/rockslicing/SeedJointSelector.scala:
```diff
@@ -0,0 +1,257 @@
+      selectedJoints
+    } else if (jointSet.length > 1) {
+      // More than 1 joint in input joint set
+      val jointOption = testVolumes(jointSet.head, jointSet.tail.head, initialBlock,
```
Let's write `jointSet(0)` and `jointSet(1)` instead of using `.head` and `.tail.head`.

shellshocked2003 replied:
I blame it all on Intellij - the highlighting when using `(0)` and recommending to use `.head` made me crack.

jhkolb commented on src/main/scala/edu/berkeley/ce/rockslicing/SeedJointSelector.scala:
```diff
@@ -0,0 +1,257 @@
+      val nonRedundantLastBlocks = lastBlocks map { case block @ Block(blockCenter, _, _) =>
+          Block(blockCenter, block.nonRedundantFaces)
+      }
+      val sortedLastBlocks = nonRedundantLastBlocks.sortWith(_.volume < _.volume)
```
All those underscores! Now that's some scala code.

jhkolb commented on src/main/scala/edu/berkeley/ce/rockslicing/SeedJointSelector.scala:
```diff
@@ -0,0 +1,257 @@
+        None
+      }
+    } else {
+      // BOTH joints intersect the block
```
Nice. I think the comments are pretty helpful in this method.

jhkolb commented on src/test/scala/edu/berkeley/ce/rockslicing/InputProcessingSpec.scala:
```diff
@@ -5,169 +5,267 @@ import org.scalatest.FunSuite
 
 class InputProcessingSpec extends FunSuite {
+
+  def compareJointSets(jointSet1: JointSet, jointSet2: JointSet): Boolean = {
```
I think, if you make the changes I've suggested above, you can compare the `JointSet`s directly with `==`. Case classes give you some handy methods for free.

shellshocked2003 replied:
I was wondering why `JointSet`s weren't working when I used `==` since they're supposed to automatically generate their own `Equals` method. Makes sense now based on your comments!

shellshocked2003 replied:
Had to add a `compareInputFaces` method for the `InputFace` case class though - the `Array` for `pointInPlane` messes up the default `Equals` method.

jhkolb commented on src/test/scala/edu/berkeley/ce/rockslicing/InputProcessingSpec.scala:
```diff
@@ -5,169 +5,267 @@ import org.scalatest.FunSuite
-      "0.0 0.0 -1.0 0.0 0.0 0.0\n" +
-      "0.0 0.0 1.0 2.0 0.0 0.0\n" +
-      "1.0 0.0 0.0 0.0 0.0 0.0 1.0 0.0 0.0 0.0 0.0\n"
+    val inputStr = "0.0 0.0 0.0 2.0 2.0 2.0\n" +
```
If you want to, you could consider creating some actual files under `src/test/resources` and then reading from them in these test cases. It might make things a bit easier to understand and maintain.

I think we have some tests somewhere that take this approach.

shellshocked2003 replied:
We use that approach in `EndToEndSpec.scala` for the input file. In this case, I think it's easier to keep these in place here: If we change the code, now we have to go back and figure out which input file corresponds to what unit test in `InputProcessingSpec` as opposed to just having it right there in front of you while trying to change/debug things.

Do you agree with that?

jhkolb replied:
Sure, we can keep it as is for now.

If the strings were a bit larger, I'd lean more strongly towards files.

jhkolb commented:
Okay, finished a second pass. Take a look.

shellshocked2003 commented:
Added case class `InputFace` for rock volume inputs. I think I addressed all your comments except for the `lazy val` and suggestion to create actual files in `/src/test/resources`. Check out my responses and let me know what you think.

jhkolb commented on src/main/scala/edu/berkeley/ce/rockslicing/InputProcessor.scala:
```diff
@@ -1,16 +1,29 @@
+  def readInput(inputSource: Source): Option[ (Array[Double], (Array[Double], Array[Double]),
+    Seq[InputFace], Seq[JointSet]) ] = {
     val lines = inputSource.getLines().zipWithIndex.toVector
     val globalOriginLine = lines.head._1
```
Can we make this `lines(0)._1` to go with the line below?

jhkolb commented on src/main/scala/edu/berkeley/ce/rockslicing/InputProcessor.scala:
```diff
@@ -24,10 +37,23 @@ object InputProcessor {
+    val boundingBoxTokens = boundingBoxLine.split(" ") map { x => Try(x.toDouble) }
+    val bbErrIndex = boundingBoxTokens.indexWhere(_.isFailure)
+    if (bbErrIndex != -1) {
+      println(s"Error, Line 2, Token $errIndex: Invalid double found in definition of bounding box")
```
Should `$errIndex` be `$bbErrIndex` on this line?

shellshocked2003 replied:
Yeah, good catch

jhkolb commented on src/main/scala/edu/berkeley/ce/rockslicing/JointGenerator.scala:
```diff
@@ -0,0 +1,210 @@
+    * @return A sequence of faces that represent the rock volume of interest
+    */
+  private def findBoundingFaces(globalOrigin: Array[Double], rockVolume: Seq[InputFace]): Seq[Face] = {
+    rockVolume map { inputFace =>
```
One thing you should consider doing is an `assert` on the length of the array arguments when you expect them to be 3D coordinates, e.g.

``` scala
assert(globalOrigin.length == 3)
```

I think I tried to add a lot of these when we switched from tuples to arrays for coordinates.

jhkolb commented on src/main/scala/edu/berkeley/ce/rockslicing/JointGenerator.scala:
```diff
@@ -0,0 +1,210 @@
+    * @return Seq of "master" joints that represent input joint sets. These will be used as a starting point
+    *         to generate full joint set.
+    */
+  private def findMasterJoints(globalOrigin: Array[Double], lowerLeftCorner: Array[Double],
```
Here's another spot where you could do an `assert` on the length of the `globalCoordinates` argument. I think there's more below, but I'll stop pointing them out now.

shellshocked2003 replied:
Good point. I'll add these where relevant.

jhkolb commented on src/main/scala/edu/berkeley/ce/rockslicing/RockSlicer.scala:
```diff
@@ -33,21 +32,34 @@ object RockSlicer {
+    val (seedJoints, remainingJoints) = if (arguments.numProcessors > 1) {
+      // Check if at least one of the input joint sets is persistent
+      if (jointSetInputs.forall(_.persistence != 100.0)) {
+        println("ERROR: Input joint sets must contain at least one persistent joint set to run in parallel. Rerun" +
```
Just curious, how often would it be the case that there's no persistent joint set?

shellshocked2003 replied:
Not very often at all for practical purposes. Also, generating a fractured rock mass from only non-persistent joints gets very tricky from an algorithm standpoint. The one we've implemented will take the first non-persistent joint that intersects the input volume and extent it through the entire rock mass. Only once the block size shrinks will non-persistence of joints feature. That's why it's important that non-persistent, geologically older joints are cut first.

Does that make sense?

jhkolb replied:
Oh yeah, I was forgetting about the age thing too.

Yep, that makes sense. Like I said, I was just curious, but I think I have a better sense now.

jhkolb commented on src/main/scala/edu/berkeley/ce/rockslicing/SeedJointSelector.scala:
```diff
@@ -0,0 +1,251 @@
+      throw new IllegalStateException("ERROR: Unable to find satisfactory seed joints")
+    }
+    if (jointSets.head(0).shape.isEmpty) {
+      // Check that joint set is persistent
```
If this comment applies to the `if` on the previous line, I think it should come on the line before the `if`. I was kind of confused if it was about the conditional or the call to `findSeedJoints`.

shellshocked2003 replied:
Sorry about that!

jhkolb commented on src/main/scala/edu/berkeley/ce/rockslicing/SeedJointSelector.scala:
```diff
@@ -0,0 +1,251 @@
+  @tailrec
+  def searchJointSets(jointSets: Seq[Seq[Joint]], inputVolume: Block, numProcessors: Int,
+                      remainingJoints: Seq[Seq[Joint]] = Seq.empty[Seq[Joint]]): (Seq[Joint], Seq[Joint]) = {
+    val allJoints = if (remainingJoints.isEmpty) {
```
I have to admit I got confused on reading over this function again. Maybe there's a better name besides `allJoints` for this value. It's also possible you could get rid of this `if` entirely...?

Intuitively, when I'm reading a recursive function, I'm trying to look at how the collections that are passed as arguments change with each call. I can understand how `jointSets` is monotonically decreasing in size since you're calling `tail` with each recursive call, but I don't understand what's happening with `remainingJoints`.

It seems to me that what you'd want to do is store all joint sets that get "passed over" here. So maybe it would be named `unusedJoints` or something like that. The collection starts out empty and, with each recursive call, you add on the joint set that you just skipped over, so your calls would look like `searchSeedJointSets(jointSets.tail, inputVolume, numProcessors, jointSets.head +: remainingJoints`.

Then, once you do find a suitable joint set, you just return `unusedJoints.flatten ++ jointSets.tail.flatten ++` whatever joints were not used as seed joints from the joint set that finally worked.

This seems like it would be a way to make the recursion clearer, but I'm probably missing something somewhere, or maybe that's what this code is actually doing and I didn't notice :)

shellshocked2003 replied:
This is essentially what the code was trying to do. I think where it got confusing is that I was just keeping the entire input joint set saved in `allJoints` -  this way `remainingJoints` was really just the entire input joint set after the first function call. Yes, this is confusing! The way you suggest returning after finding a suitable joint set is what the function was doing, but I think because of the confusion about `remainingJoints` it wasn't clear.
I've updated the function so that the `if` at the beginning is gone and that `remainingJoints` is actually just the remaining joints. `remainingJoints` and `allOtherJoints` - which is what is returned - maintain the ordering of the joint sets in the input joint set.

jhkolb commented on src/test/scala/edu/berkeley/ce/rockslicing/BlockVTKSpec.scala:
```diff
@@ -163,12 +165,12 @@ class BlockVTKSpec extends FunSuite {
-      face4 -> Seq[Int](7, 4, 3, 2),
-      face5 -> Seq[Int](2, 0, 6, 7),
-      face6 -> Seq[Int](4, 5, 1, 3)
+      face1 -> Seq[Int](7, 1, 3, 6),
```
All these changes look annoying. Just curious, is there anything we can do to make these unit tests less brittle?

Obviously this wouldn't be something to add to this pull request, just wanted to hear your thoughts on this.

shellshocked2003 replied:
These changes are definitely annoying. I'm not sure about how to make these less brittle. The thing that messes it up is that everything is based on the order in which the vertices appear in the array of vertices. If this changes, the unit tests change too. 
I'm not sure of a practical way to ensure that the vertices are always stored starting from the same one. Do you have any ideas?

jhkolb replied:
Not really. I'll add it as an issue and we can leave it alone for now.

jhkolb commented on src/test/scala/edu/berkeley/ce/rockslicing/InputProcessingSpec.scala:
```diff
@@ -5,169 +5,261 @@ import org.scalatest.FunSuite
 
 class InputProcessingSpec extends FunSuite {
+
+  def compareInputFaces(face1: InputFace, face2: InputFace): Boolean = {
```
How about simplifying the body of this function to:

``` scala
face1.strike == face2.strike &&
       face1.dip == face2.dip &&
       (face1.pointInPlane sameElements face2.pointInPlane) &&
       face1.phi == face2.phi &&
       face1.cohesion == face2.cohesion
```

Should avoid the unnecessary `if` and have the same effect.

Also, change the name to show that it's an equality check and not a comparison.

You may also want to consider just overriding the `equals` method in the `InputFace` case class.

shellshocked2003 replied:
This was pretty messy, I've added an `equals` method instead.

jhkolb commented on src/test/scala/edu/berkeley/ce/rockslicing/JointGeneratorSpec.scala:
```diff
@@ -0,0 +1,347 @@
+        compareJoints(joint1, joint2)
+    }
+
+    assert(!jointComparison.contains(false))
```
You can simplify this to

``` scala
assert(expectedJoints.zip(generatedInput.masterJoints).forAll { case (joint1, joint2) => compareJoints(joint1, joint2) })
```

At least in my experience, a lot of times when you have an assertion or a return involving a literal `true` or `false`, the code is actually more complicated than it needs to be.

Also, we shouldn't use the name `compareJoints`, which implies checking two joints for the relative ordering under some scheme, when this is really an equality check.

shellshocked2003 replied:
See comment below

jhkolb commented on src/test/scala/edu/berkeley/ce/rockslicing/JointGeneratorSpec.scala:
```diff
@@ -0,0 +1,347 @@
+    assert((expectedOrigin sameElements generatedInput.origin) &&
+      (expectedLowerLeftCorner sameElements generatedInput.lowerLeftCorner) &&
+      (expectedUpperRightCorner sameElements generatedInput.upperRightCorner) &&
+      !jointComparison.contains(false))
```
See my comment above, this is a similar case where you should use a `forAll`.

jhkolb commented on src/test/scala/edu/berkeley/ce/rockslicing/JointGeneratorSpec.scala:
```diff
@@ -0,0 +1,347 @@
+    assert((expectedOrigin sameElements generatedInput.origin) &&
+      (expectedLowerLeftCorner sameElements generatedInput.lowerLeftCorner) &&
+      (expectedUpperRightCorner sameElements generatedInput.upperRightCorner) &&
+      !jointComparison.contains(false))
```
Use `forAll`

jhkolb commented on src/test/scala/edu/berkeley/ce/rockslicing/JointGeneratorSpec.scala:
```diff
@@ -0,0 +1,347 @@
+    assert((expectedOrigin sameElements generatedInput.origin) &&
+      (expectedLowerLeftCorner sameElements generatedInput.lowerLeftCorner) &&
+      (expectedUpperRightCorner sameElements generatedInput.upperRightCorner) &&
+      !jointComparison.contains(false))
```
Use `forAll`

jhkolb commented on src/test/scala/edu/berkeley/ce/rockslicing/SeedJointSelectorSpec.scala:
```diff
@@ -0,0 +1,212 @@
+
+class SeedJointSelectorSpec extends FunSuite {
+
+  private def compareJoints(joint1: Joint, joint2: Joint): Boolean = {
```
You should name this something different to indicate that it's an equality check and not a comparison.

This looks like it's the exact same logic as the other `compareJoints` function right? Seems kind of silly to implement the same thing twice, so perhaps we need to add something to the `Joint` class, if our current `equals` method isn't suitable.

shellshocked2003 replied:
I added an `approximateEquals` method to the `Joint` and used that in the unit tests instead.

jhkolb commented on src/test/scala/edu/berkeley/ce/rockslicing/SeedJointSelectorSpec.scala:
```diff
@@ -0,0 +1,212 @@
+      compareJoints(joint1, joint2)
+    }
+
+    assert(!jointComparison.contains(false))
```
Drop the explicit `val` above and use a `forAll` in the assert.

jhkolb commented on src/test/scala/edu/berkeley/ce/rockslicing/SeedJointSelectorSpec.scala:
```diff
@@ -0,0 +1,212 @@
+      compareJoints(joint1, joint2)
+    }
+
+    assert(!jointComparison.contains(false))
```
Use `forAll`

jhkolb commented on src/test/scala/edu/berkeley/ce/rockslicing/SeedJointSelectorSpec.scala:
```diff
@@ -0,0 +1,212 @@
+      compareJoints(joint1, joint2)
+    }
+
+    assert(!jointComparison.contains(false))
```
Use `forAll`

jhkolb commented on src/test/scala/edu/berkeley/ce/rockslicing/SeedJointSelectorSpec.scala:
```diff
@@ -0,0 +1,212 @@
+      compareJoints(joint1, joint2)
+    }
+
+    assert(!jointComparison.contains(false))
```
Use `forAll`

jhkolb commented on src/test/scala/edu/berkeley/ce/rockslicing/JointGeneratorSpec.scala:
```diff
@@ -0,0 +1,347 @@
+
+class JointGeneratorSpec extends FunSuite {
+
+  private def compareJoints(joint1: Joint, joint2: Joint): Boolean = {
```
Name this something different to indicate that it's an equality check and not a comparison.

Also, I thought we already had a nice `equals` method in `Joint.scala`? Or did we have an `approximateEquals` method at some point too?

jhkolb commented:
Few more things. We're getting close though!

shellshocked2003 commented:
`forall` for all :) Also, updated `searchJointSets` so it's clearer how the recursion is working.

jhkolb commented on src/main/scala/edu/berkeley/ce/rockslicing/Joint.scala:
```diff
@@ -314,29 +312,27 @@ case class Joint(normalVec: Array[Double], localOrigin: Array[Double],
-       (math.abs(face.c + c) < NumericUtils.EPSILON) &&
-       (math.abs(face.d + distance) < NumericUtils.EPSILON))
-    }
+  def approximateEquals(inputJoint: Joint, tolerance: Double = NumericUtils.EPSILON): Boolean = {
```
Looks nice!

jhkolb commented on src/main/scala/edu/berkeley/ce/rockslicing/JointGenerator.scala:
```diff
@@ -0,0 +1,219 @@
+  */
+case class JointGenerator(globalOrigin: Array[Double], boundingBox: (Array[Double], Array[Double]),
+                          rockVolumeInputs: Seq[InputFace], jointSetData: Seq[JointSet]) {
+  assert(globalOrigin.length == 3)
```
Cool. These `asserts` are nice because if something weird starts happening, we'll know about it.

jhkolb commented on src/main/scala/edu/berkeley/ce/rockslicing/SeedJointSelector.scala:
```diff
@@ -0,0 +1,244 @@
+    *         joint set.
+    */
+  @tailrec
+  def searchJointSets(jointSets: Seq[Seq[Joint]], inputVolume: Block, numProcessors: Int,
```
This makes a lot more sense now. :)

jhkolb commented:
Looks good. Merging.
